lua_gettable's return type is int, not void in lua-5.3

conky-1.10.0 fails to compile against lua-5.3; the error is:


In file included from /usr/include/lua.hpp:6:0,
                 from /build/conky/src/conky-1.10.0/src/luamm.hh:30,
                 from /build/conky/src/conky-1.10.0/src/luamm.cc:24:
/build/conky/src/conky-1.10.0/src/luamm.cc: In member function 'void lua::state::gettable(int)':
/build/conky/src/conky-1.10.0/src/luamm.cc:370:3: error: no matches converting function 'safe_misc_trampoline' to type 'lua_CFunction {aka int (*)(struct lua_State*)}'
   lua_pushcfunction(cobj.get(), (&safe_misc_trampoline<&lua_gettable, 1>));
   ^
/build/conky/src/conky-1.10.0/src/luamm.cc:168:7: note: candidate is: template<void (* misc)(lua_State*, int), int nresults> int lua::{anonymous}::safe_misc_trampoline(lua_State*)
   int safe_misc_trampoline(lua_State *l)
       ^
src/CMakeFiles/conky.dir/build.make:675: recipe for target 'src/CMakeFiles/conky.dir/luamm.cc.o' failed


It turns out lua_gettable returns an int in lua-5.3. The attached patch fixes this issue, though as currently written it is specific to lua-5.3.

--- a/src/luamm.cc
+++ b/src/luamm.cc
@@ -171,6 +171,14 @@ namespace lua {
 			return nresults;
 		}
 
+		// Overloaded for Lua 5.3+ as lua_gettable and others return an int
+		template<int (*misc)(lua_State *, int), int nresults>
+		int safe_misc_trampoline(lua_State *l)
+		{
+			misc(l, 1);
+			return nresults;
+		}
+
 		int safe_next_trampoline(lua_State *l)
 		{
 			int r = lua_next(l, 1);

